tests/test_mutations.py

```diff
@@ -472,45 +779,308 @@ def test_keep_mutation_parent_zero_rate(self):
+                    assert position < edge.right
+                    if position not in self.sites:
+                        site = self.add_site(
+                                        position=position,
+                                        new=True,
+                                        ancestral_state="X",
+                                        metadata=b'')
+                    time = self.rng.flat(branch_start, branch_end)
+                    self.add_mutation(
+                            id=tskit.NULL,
+                            site=site,
+                            node=edge.child,
+                            time=time,
+                            new=True,
+                            derived_state="X",
+                            metadata=b"")
+                index += 1
+                left = right
+
+    def pick_allele(self):
+        u = self.rng.flat(0, 1)
+        j = 0
+        while u > self.alphabet_distribution[j]:
+            u -= self.alphabet_distribution[j]
+            j += 1
+        return j
+
+    def transition_allele(self, j):
+        u = self.rng.flat(0, 1)
+        k = 0
+        while u > self.mutation_matrix[j][k]:
+            u -= self.mutation_matrix[j][k]
+            k += 1
+        return k
+
+    def choose_alleles(self, tree_parent, site, mutation_id):
+        if site.new:
+            ai = self.pick_allele()
+            site.ancestral_state = self.alleles[ai]
+        else:
+            ai = self.alleles.index(site.ancestral_state)
+        # sort mutations by (increasing id if both are not null,
+        #  decreasing time, increasing insertion order)
+        print([s.id for s in site.mutations])
+        site.mutations.sort(key=functools.cmp_to_key(cmp_mutation))
+        print([s.id for s in site.mutations])
+        bottom_mutation = {}
+        j = 0
+        while j < len(site.mutations):
+            mut = site.mutations[j]
+            mut.id = mutation_id + j
+            # Traverse up the tree to find the parent mutation
+            # bottom_mutation[u] is the index in mutations of the most recent
+            #    mutation seen on the edge above u so far, if any
+            u = mut.node
+            while u != tskit.NULL and u not in bottom_mutation:
+                u = tree_parent[u]
+            if u == tskit.NULL:
+                # parent is the ancestral state
+                pa = ai
+                assert mut.parent == tskit.NULL
+            else:
+                assert u in bottom_mutation
+                parent_mut = site.mutations[bottom_mutation[u]]
+                mut.parent = parent_mut.id
+                assert mut.parent < mut.id, "Parent after child mutation."
+                assert mut.time <= parent_mut.time, "Parent after child mutation."
+                if mut.new:
+                    pa = self.alleles.index(parent_mut.derived_state)
+                else:
+                    # check that no kept mutation occurs after a new mutation
```

Comment:
This would result from bad input - e.g. generating mutations twice on the same stretch of time, or generating mutations on a more recent time chunk before a more distant time chunk.

Reply:
I mean, at this point logic error is more likely, but when it all works, it'll be due to bad input. =)